mne/preprocessing/xdawn.py

```diff
@@ -1,266 +1,362 @@
-        The data to estimates evoked
-    events : ndarray, shape (n_events, 3)
+    epochs_data : array, shape (n_channels, n_times)
+        The epochs data to estimate evoked.
+    events : array, shape (n_events, 3)
         The events typically returned by the read_events function.
         If some events don't match the events of interest as specified
         by event_id, they will be ignored.
-    event_id : dict
-        The id of the events to consider
     tmin : float
         Start time before event.
-    tmax : float
-        End time after event.
     sfreq : float
         Sampling frequency.
 
     Returns
     -------
-    evokeds_data : dict of ndarray
-        A dict of evoked data for each event type in event_id.
-    toeplitz : dict of ndarray
-        A dict of toeplitz matrix for each event type in event_id.
+    evokeds_data : array, shape (n_class * n_components, n_channels)
+        An concatenated array of evoked data for each event type.
+    toeplitz : array, shape (n_class * n_components, n_channels)
+        An concatenated array of toeplitz matrix for each event type.
     """
-    nmin = int(tmin * sfreq)
-    nmax = int(tmax * sfreq)
-
-    window = nmax - nmin
-    n_samples = data.shape[1]
-    toeplitz_mat = dict()
-    full_toep = list()
-    for eid in event_id:
+
+    n_epochs, n_channels, n_times = epochs_data.shape
+    tmax = tmin + n_times / float(sfreq)
+
+    # Deal with shuffled epochs
+    events = events.copy()
+    events[:, 0] -= events[0, 0] + int(tmin * sfreq)
+
+    # Contruct raw signal
+    raw = _construct_signal_from_epochs(epochs_data, events, sfreq, tmin)
+
+    # Compute average evoked
+    n_min, n_max = int(tmin * sfreq), int(tmax * sfreq)
+    window = n_max - n_min
+    n_samples = raw.shape[1]
+    toeplitz = list()
+    classes = np.unique(events[:, 2])
+    for ii, this_class in enumerate(classes):
         # select events by type
-        ix_ev = events[:, -1] == event_id[eid]
+        sel = events[:, 2] == this_class
 
         # build toeplitz matrix
         trig = np.zeros((n_samples, 1))
-        ix_trig = (events[ix_ev, 0]) + nmin
+        ix_trig = (events[sel, 0]) + n_min
         trig[ix_trig] = 1
-        toep_mat = linalg.toeplitz(trig[0:window], trig)
-        toeplitz_mat[eid] = toep_mat
-        full_toep.append(toep_mat)
+        toeplitz.append(linalg.toeplitz(trig[0:window], trig))
 
     # Concatenate toeplitz
-    full_toep = np.concatenate(full_toep)
+    toeplitz = np.array(toeplitz)
+    full_toep = np.concatenate(toeplitz)
 
     # least square estimation
     predictor = np.dot(linalg.pinv(np.dot(full_toep, full_toep.T)), full_toep)
-    all_evokeds = np.dot(predictor, data.T)
-    all_evokeds = np.vsplit(all_evokeds, len(event_id))
-
-    # parse evoked response
-    evoked_data = dict()
-    for idx, eid in enumerate(event_id):
-        evoked_data[eid] = all_evokeds[idx].T
+    evokeds_data = np.dot(predictor, raw.T)
```

Comment:
Look here: it *is* X, this is simply constructing the pseudo inverse, and solving for the coefs in line 97.

It's literally just a least squares estimate.

```
pinv = np.dot(linalg.pinv(np.dot(X, X.T)), X)
np.dot(pinv, y)
```